[CONTINT-4562] Container generator

L3n41c opened the pull request:
### What does this PR do?

A form of load can be a container that exists and exposes some data.

This generator aims to provide a way to start a generic container based on user's config.

### Motivation

Test the agent performances when it monitors many containers

### Related issues

### Additional Notes

Follow-up of #1154.

blt reviewed (commented):
I really like where this is headed.

blt commented on lading/src/generator/container.rs:
```diff
@@ -0,0 +1,245 @@
+
+        for _ in 0..self.nb {
+            let container_name = format!("lading_container_{}", Uuid::new_v4());
+            info!("Creating container: {}", container_name);
```
This could be slightly changed to `info!("Creating container: {container_name}");`. String formatting allows for variable names to be present within `{}`.

L3n41c replied:
Fixed in https://github.com/DataDog/lading/pull/1227/commits/59d040caefd33fe0de219df051dbfd89d3dce03c.

blt commented on lading/src/generator/container.rs:
```diff
@@ -0,0 +1,245 @@
+                )
+                .await?;
+
+            info!("Created container with id: {}", container.id);
```
REF the above string formatting comment, if you wanted to do the same thing here you'd have to define `container.id` a name, so like `info!("Created container with id: {id}", id = container.id);` More an FYI than anything. It can be helpful when there are a lot of fields to keep track of.

L3n41c replied:
👍🏻

blt commented on lading/src/generator/container.rs:
```diff
@@ -0,0 +1,245 @@
+        let docker = Docker::connect_with_local_defaults()?;
+
+        let full_image = format!("{}:{}", self.image, self.tag);
+        info!("Ensuring image is available: {}", full_image);
```
This and the below logs, could we bump these down to debug? In practice we'll see tons of these logs and they strike me as more useful in a debugging scenario. CC @cmetz100, I'm curious also for your thoughts. I know log volume in our system is an area you think on a fair bit.

L3n41c replied:
I switched all the `info` logs to `debug` except the very first `Container generator running …` one in https://github.com/DataDog/lading/pull/1227/commits/59d040caefd33fe0de219df051dbfd89d3dce03c.

blt commented on lading/src/generator/container.rs:
```diff
@@ -0,0 +1,245 @@
+    labels: Option<HashMap<String, String>>,
+    network_disabled: Option<bool>,
+    exposed_ports: Option<Vec<String>>,
+    nb: usize,
```
Ah yeah, following on from the comment on the Config field, `nb` as a field name here gives me very little understanding of what it is for.

L3n41c replied:
I renamed it `number_of_containers` in https://github.com/DataDog/lading/pull/1227/commits/59d040caefd33fe0de219df051dbfd89d3dce03c.

blt commented on lading/src/generator/container.rs:
```diff
@@ -0,0 +1,245 @@
+    /// Ports to expose from the container
+    pub exposed_ports: Option<Vec<String>>,
+    /// Number of containers to spin up (defaults to 1)
+    pub nb_containers: Option<u32>,
```
Looking at this I'm not totally sure what `nb` means. I think probably "number"? In general lading house style is to spell things out, so probably `total_containers` and further adjustment of `nb` below is in keeping with that.

L3n41c replied:
I renamed it `number_of_containers` in https://github.com/DataDog/lading/pull/1227/commits/59d040caefd33fe0de219df051dbfd89d3dce03c.

blt reviewed (commented):
Some style comments on the commits. I would like to clarify what happens if the spawned container exits. As I read it the `Container` generator will not check for spawned container liveness and so if a spawned container fails the user won't know until `stop_and_remove_container` is called? There's potentially a long period of time between startup and shutdown so an earlier warning -- of outright failure of lading -- is desirable. I could buy may-fail or must-not-fail approaches to this generator, but documenting which approach is taken here is helpful to the user.

L3n41c commented on lading/src/generator.rs:
```diff
@@ -239,6 +250,8 @@ impl Server {
             Server::PassthruFile(inner) => inner.spin().await?,
             Server::ProcessTree(inner) => inner.spin().await?,
             Server::ProcFs(inner) => inner.spin().await?,
+            // Run the container generator
```
Agreed! Fixed in https://github.com/DataDog/lading/pull/1227/commits/59d040caefd33fe0de219df051dbfd89d3dce03c.

L3n41c commented on lading/src/generator/container.rs:
```diff
@@ -0,0 +1,245 @@
+    pub repository: String,
+    /// The container image tag (e.g. "latest")
+    pub tag: String,
+    /// Arguments to provide to the container (docker calls this args, but that's a dumb name)
```
Agreed! Fixed in https://github.com/DataDog/lading/pull/1227/commits/59d040caefd33fe0de219df051dbfd89d3dce03c.

L3n41c commented:
@blt Thank you for your review!
I should have addressed all the comments.

> I would like to clarify what happens if the spawned container exits. As I read it the Container generator will not check for spawned container liveness and so if a spawned container fails the user won't know until stop_and_remove_container is called? There's potentially a long period of time between startup and shutdown so an earlier warning -- of outright failure of lading -- is desirable.

I’ve also updated the PR so that `lading` is now periodically checking if the spawned containers are still running. If they’re not, a warning is issued.

blt commented on lading/src/generator/container.rs:
```diff
@@ -0,0 +1,251 @@
+        loop {
+            tokio::select! {
+                // Check that containers are still running every 10 seconds
+                () = tokio::time::sleep(tokio::time::Duration::from_secs(10)) => {
```
This would be better done in terms of `tokio::time::interval`, example [here](https://github.com/DataDog/lading/blob/68b7642e0c7dc59b69559aa43f5ddb1329cb0a57/lading/src/bin/lading.rs#L529).

blt reviewed (commented):
One small comment about using interval instead of repeatedly creating sleeps, otherwise looks good. I'll be happy to see this merge up after that's fixed.
